DataFormats/Provenance/interface/IndexIntoFile.h

```diff
@@ -49,12 +49,34 @@ There can be multiple run entries in a TTree associated
 with the same run number and ProcessHistoryID in a file.
 There can also be multiple lumi entries associated with
 the same lumi number, run number, and ProcessHistoryID.
-Both sorting orders will make these subgroups contiguous,
+The sorting orders identified as numericalOrder and
+firstAppearanceOrder will make these subgroups contiguous,
 but beyond that is up to the client (normally PoolSource,
 which passes them up to the EventProcessor)
 to deal with merging the multiple run (or lumi) entries
 together.
 
+In entryOrder, the Events are processed in the order they
+appear in the Events TTree. The event entries associated
+with a run might not be contiguous. The event entries associated
+with a lumi might not be contiguous. Even so, the iteration
+order will always follow the pattern run followed by contained
+lumis followed by contained events (note it's possible
+there are no contained events for a lumi or no contained
+lumis for a run). Because the events are not contiguous,
+this leads to run or lumi entries appearing more than once
+in the iteration (once for each contained contiguous sequence
+of events plus possibly one additional time near the end
+for the run or lumi to be processed and merged, although this
+might or might not occur with the last contiguous sequence
+of events). The iterator automatically skips over
+the duplicate event entries created by this. The client
+needs to check the iterator function named shouldProcessLumi
+or shouldProcessRun to determine whether a lumi or run
+entry should be recognized as a placeholder or it is
+the entry where the products should be read and merged
+for output.
```

Comment:
The transitions will occur every time the iterator hits a run or lumi entry in the iteration. But PoolSource will not read the run or lumi products or call writeRun/writeLumi unless shouldProcessRun or shouldProcessLumi return true. When they are false, any attempt by a module to get a run or lumi product results in a 'ProductNotFound' exception or invalid Handle. (There is an exceptional tricky case when crossing file boundaries and a valid one gets combined with an invalid one)

Previously, we just said you are not allowed to use modules that access run or lumi products in noRunLumiSort mode and I am assuming we will continue that.

I considered adding a function to the Run class (or LuminosityBlock class) providing this information. The Principal knows. Then a module could check and when true actually use run or lumi products in beginRun/endRun/beginLumi/endLumi transitions. It still would not work in events. I didn't do this because it is not required for the merging use case and it is a little complicated. It would be trivial to add though. 

This behavior results in the same merged run or lumi products in output files that we get with the other two iterator types.